File: app/src/main/java/org/standardserve/driftflasche/dialog/MarkerCreationDialog.java
package org.standardserve.driftflasche.dialog;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Looper;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;

import com.google.android.gms.maps.GoogleMap;
import com.google.android.material.button.MaterialButtonToggleGroup;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.android.material.textfield.TextInputLayout;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.FormBody;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

import org.json.JSONException;
import org.json.JSONObject;
import org.standardserve.driftflasche.R;
import org.standardserve.driftflasche.network.NetworkAccess;

import java.io.IOException;
import java.util.Objects;

// Bottle types
enum BOTTLE_TYPE{
    INFO, MOOD, WARN
}

// Dialog for creating a new marker
public class MarkerCreationDialog {
    private MarkerCreationDialog(){}

    private static BOTTLE_TYPE bootleType = BOTTLE_TYPE.INFO; // 1: info, 2: mood, 3: warn
    private static String bootleContent = ""; // Content of the bottle

    // toogle selection initialization
    @SuppressLint("NonConstantResourceId")
    private static void setToogleButton(MaterialButtonToggleGroup toogleButton){
        // Set the toogle button
        toogleButton.addOnButtonCheckedListener(
                (group, checkedId, isChecked) -> {
                    if(isChecked){
                        switch (checkedId){
                            case R.id.info: bootleType = BOTTLE_TYPE.INFO; break;
                            case R.id.mood: bootleType = BOTTLE_TYPE.MOOD; break;
                            case R.id.warn: bootleType = BOTTLE_TYPE.WARN; break;
                            default: break;
                        }
                    }
                }
        );
    }

    // Create the dialog
    public static void create(Context context, String username, Double lat, Double lng, String token, GoogleMap map) {
        LayoutInflater inflater = LayoutInflater.from(context); // Get the layout inflater
        View view = inflater.inflate(R.layout.set_bottle, null); // Inflate the layout
        MaterialButtonToggleGroup toogleButton = view.findViewById(R.id.bottleType); // Get the toogle button
        TextInputLayout dialogText = view.findViewById(R.id.comment); // Get the text input
        setToogleButton(toogleButton); // Set the toogle button
        AlertDialog builder = new MaterialAlertDialogBuilder(context)
                .setTitle(R.string.bottle_creation_dialog_title)
                .setPositiveButton(R.string.confirm, (dialog, which) -> {
                    // Get the content of the text input
                    OkHttpClient mOKHttpClient = new OkHttpClient();
                    bootleContent = Objects.requireNonNull(dialogText.getEditText()).getText().toString();
                    RequestBody formBody = new FormBody.Builder()
                            .add("username", username)
                            .add("lat", lat.toString())
                            .add("lng", lng.toString())
                            .add("type", bootleType.toString())
                            .add("content", bootleContent)
                            .add("token", token)
                            .add("mode", "add")
                            .build();
                    String url = NetworkAccess.getBOTTLEAccess();
                    Request request = new Request.Builder()
                            .url(url)
                            .post(formBody)
                            .build();
                    Call call = mOKHttpClient.newCall(request);
                    call.enqueue(new Callback() {
                        @Override
                        public void onFailure(@NonNull Call call, @NonNull IOException e) {
                            Looper.prepare();
                            Toast.makeText(context, "Login failed due to request error", Toast.LENGTH_SHORT).show();
                            Looper.loop();
                        }

                        @Override
                        public void onResponse(@NonNull Call call, @NonNull Response response) throws IOException {
                            // Get the response of bottle creation
                            String str = Objects.requireNonNull(response.body()).string();
                            if(str.length()>0) {
                                JSONObject receiveObj = null;
                                try {
                                    receiveObj = new JSONObject(str);
                                } catch (JSONException e) {
                                    e.printStackTrace();
                                    Looper.prepare();
                                    Toast.makeText(context, "Login failed due to request error", Toast.LENGTH_SHORT).show();
                                    Looper.loop();
                                }
                                String status = null;
                                String msg = null;
                                try {
                                    assert receiveObj != null;
                                    status = receiveObj.getString("status");
                                    msg = receiveObj.getString("msg");
                                } catch (JSONException e) {
                                    e.printStackTrace();
                                    Looper.prepare();
                                    Toast.makeText(context, "Login failed due to request error", Toast.LENGTH_SHORT).show();
                                    Looper.loop();
                                }
                                assert status != null;
                                if(Integer.parseInt(status) == 0){
                                    // refresh the map if the bottle is created successfully
                                    bottlesReload.loadBottlesbyDistance(context, 20, lat, lng, token, username, map, "");
                                }
                                Looper.prepare();
                                Toast.makeText(context, msg, Toast.LENGTH_SHORT).show();
                                Looper.loop();
                            }else{
                                Looper.prepare();
                                Toast.makeText(context, "Login failed due to request error", Toast.LENGTH_SHORT).show();
                                Looper.loop();
                            }
                        }
                    });

                })
                .setNegativeButton(R.string.decline, (dialog, which) -> dialog.dismiss())
                .setView(view)
                .create();
        builder.show();
    }
}

File: app/src/main/java/org/standardserve/driftflasche/dialog/MarkerCreationDialog.kt
package org.standardserve.driftflasche.dialog

import android.content.Context
import android.content.DialogInterface
import android.content.res.Resources
import android.os.Looper
import android.util.Log
import android.view.LayoutInflater
import android.widget.Toast
import androidx.core.widget.addTextChangedListener
import com.google.android.material.button.MaterialButtonToggleGroup
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.textfield.TextInputEditText
import okhttp3.FormBody
import okhttp3.MediaType
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import org.standardserve.driftflasche.R
import org.standardserve.googletestunit.fileio.RootPath
import org.standardserve.googletestunit.fileio.TokenReadAndWrite
import kotlin.concurrent.thread

enum class BOOTLE_TYPE{
    INFO, MOOD, WARN
}
object MarkerCreationDialog {

    private var bootleType = BOOTLE_TYPE.INFO // 1: info, 2: mood, 3: warn
    private var bootleContent = ""

    private fun setToogleButton(toogleButton: MaterialButtonToggleGroup){
        toogleButton.addOnButtonCheckedListener { group, checkedId, isChecked ->
            if (isChecked) {
                when(checkedId) {
                    R.id.info -> {
                        bootleType = BOOTLE_TYPE.INFO
                    }
                    R.id.mood -> {
                        bootleType = BOOTLE_TYPE.MOOD
                    }
                    R.id.warn -> {
                        bootleType = BOOTLE_TYPE.WARN
                    }
                    else -> {
                        bootleType = BOOTLE_TYPE.INFO
                    }
                }
            }
        }
    }

    private fun setDialogText(textView: TextInputEditText){
        textView.addTextChangedListener {
            bootleContent = it.toString()
        }
    }

    fun create(context: Context, username: String, lat: Double, lng: Double, token: String) {
        val inflater = LayoutInflater.from(context)
        val view = inflater.inflate(R.layout.set_bottle, null)
        var toogleButton = view.findViewById<com.google.android.material.button.MaterialButtonToggleGroup>(R.id.bottleType)
        var dialogText = view.findViewById<com.google.android.material.textfield.TextInputEditText>(R.id.comment)
        setToogleButton(toogleButton)
        setDialogText(dialogText)
        val dialog = MaterialAlertDialogBuilder(context)
            .setTitle(R.string.bottle_creation_dialog_title).
            setPositiveButton(R.string.confirm) { dialog, which ->
                // Respond to neutral button press
                var url = "http://138.68.65.184:5000/api/login"
                var formData = JSONObject()
                formData.put("mode", "add")
                formData.put("type", bootleType)
                formData.put("username", username)
                formData.put("content", bootleContent)
                formData.put("lat", lat)
                formData.put("lng", lng)
                var strForm = formData.toString()
                val requestBody = strForm?.let {
                    val contentType: MediaType = "application/json".toMediaType()
                    strForm.toRequestBody(contentType)
                } ?: run {
                    FormBody.Builder().build()
                }
                thread {
                    var client = OkHttpClient()
                    var request = Request.Builder()
                        .url(url)
                        .post(requestBody)
                        .build()
                    var response = client.newCall(request).execute()
                    var resData = response.body?.string()
                    if(resData != null) {
                        val receiveObj = JSONObject(resData)
                        val status = receiveObj.getString("status")
                        val msg = receiveObj.getString("msg")
                        if (status.toInt() == 0) {
                            Looper.prepare()
                            Toast.makeText(context, msg, Toast.LENGTH_SHORT).show()
                            Looper.loop()
                        }else{
                            Looper.prepare()
                            Toast.makeText(context, msg, Toast.LENGTH_SHORT).show()
                            Looper.loop()
                        }


                    }else{
                        Looper.prepare()
                        Toast.makeText(context, "Login failed due to request error", Toast.LENGTH_SHORT).show()
                        Looper.loop()
                    }
                }

            }
            .setNegativeButton(R.string.decline) { dialog, which ->
                // Respond to negative button press
            }
            .setView(view)
            .create()
        dialog.show()

    }
}
